### PR TITLE
feat: [CDS-56763]: Modify all requests/responses for preference api to include a new field for usergroups. 

### PR DESCRIPTION
**Pull request by rafaelmart1ns**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/45375)

### CODE REVIEW

**Comment by rafaelmart1ns** [comment]
trigger messagecheck

**Comment by rafaelmart1ns** [comment]
trigger feature-build

**Comment by rafaelmart1ns** [comment]
trigger ut6

File: 400-rest/src/main/java/software/wings/service/impl/PreferenceServiceImpl.java [hunk 1]
```diff
@@ -41,14 +49,19 @@
...
 import java.util.Set;
+import java.util.stream.Collectors;
 import lombok.extern.slf4j.Slf4j;
 
 @Singleton
 @Slf4j
 public class PreferenceServiceImpl implements PreferenceService {
-  public static final String USER_ID_KEY = "userId";
   public static final String PREFERENCE_WITH_SAME_NAME_EXISTS = "Preference with same name exists for user";
+  public static final String USER_IS_NOT_ENABLE = "User is not a enable to add a share filter";
```
Comment (lasdox) [inline 400-rest/src/main/java/software/wings/service/impl/PreferenceServiceImpl.java:59]: public static final String USER_IS_NOT_ENABLED = "User is not enabled to add a share filter";

**Review by rafaelmart1ns — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45375)* status: 9 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/PreferenceServiceImpl.java` line 59 at r5](https://reviewable.io/reviews/harness/harness-core/45375#-NR-dZew-7WxMuzE2Pco-r5-59:-NR9wghTBPg1z4ai1AFc:bozgwtm) ([raw file](https://github.com/harness/harness-core/blob/a19872262520cf2205b7c7ebca277045624ee4b6/400-rest/src/main/java/software/wings/service/impl/PreferenceServiceImpl.java#L59)):*
<details><summary><i>Previously, lasdox (Lucas Sales) wrote…</i></summary><blockquote>

public static final String USER_IS_NOT_ENABLED = "User is not enabled to add a share filter";
</blockquote></details>

it changed.

File: 400-rest/src/main/java/software/wings/service/impl/PreferenceServiceImpl.java [hunk 2]
```diff
@@ -58,25 +71,45 @@ public Preference save(String accountId, String userId, Preference preference) {
...
 
   @Override
   public Preference get(String accountId, String userId, String preferenceId) {
     Preference preference = wingsPersistence.createQuery(Preference.class)
-                                .filter(PreferenceKeys.accountId, accountId)
-                                .filter(USER_ID_KEY, userId)
-                                .filter(PreferenceKeys.uuid, preferenceId)
+                                .filter(PreferenceKeys.ACCOUNT_ID, accountId)
+                                .filter(PreferenceKeys.USER_ID, userId)
+                                .filter(PreferenceKeys.ID, preferenceId)
```
Comment (lasdox) [inline 400-rest/src/main/java/software/wings/service/impl/PreferenceServiceImpl.java:112]: Why these changes are needed?
defaut convention for CollectionKeys is camelCase

Comment (rafaelmart1ns) [reply 400-rest/src/main/java/software/wings/service/impl/PreferenceServiceImpl.java:112]: The sonarlint added this as the right way to declare that and it fails in ci if using another format.

Comment (rafaelmart1ns) [reply 400-rest/src/main/java/software/wings/service/impl/PreferenceServiceImpl.java:112]: It changed.

File: 400-rest/src/main/java/software/wings/service/impl/PreferenceServiceImpl.java [hunk 3]
```diff
@@ -87,8 +120,29 @@ public Preference get(String accountId, String userId, String preferenceId) {
...
+    if (featureFlagService.isEnabled(SPG_ENABLE_SHARING_FILTERS, accountId)) {
+      SearchFilter userIdFilter =
+          SearchFilter.builder().fieldName(PreferenceKeys.USER_ID).fieldValues(new String[] {userId}).op(EQ).build();
+      for (String s : l) {
+        pageRequest.addFilter("", SearchFilter.Operator.OR,
+            SearchFilter.builder()
+                .fieldName(PreferenceKeys.USER_GROUPS_ID_TO_SHARE)
+                .op(SearchFilter.Operator.EQ)
+                .fieldValues(new String[] {s})
+                .build(),
```
[the remark this answers is not in the document]
Comment (rafaelmart1ns) [reply 400-rest/src/main/java/software/wings/service/impl/PreferenceServiceImpl.java:139]: because I don't know why the PageRequest filter support both operations in the same filter.

**Review by lasdox — APPROVED** [review]
> Reviewed 1 of 8 files at r3, 3 of 8 files at r4, 1 of 2 files at r5, 1 of 2 files at r6, 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45375)* status: 9 file(s) to review. 1 of 2 approvals obtained.

**Comment by rafaelmart1ns** [comment]
trigger utAll

**Review by rafaelmart1ns — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45375#-:-NRJL8DL7ZHlPPf7T4Jn:bnfp4nl)* status: 1 file(s) to review.

**Review by rafaelmart1ns — COMMENTED** [review]
> Reviewed 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45375)* status: 1 file(s) to review.

**Comment by rafaelmart1ns** [comment]
trigger ut2

File: 400-rest/src/main/java/software/wings/service/impl/PreferenceServiceImpl.java (hunk 1 above)
Comment (rafaelmart1ns) [reply 400-rest/src/main/java/software/wings/service/impl/PreferenceServiceImpl.java]: it changed

**Review by lasdox — APPROVED** [review]
> Reviewed 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45375)* status: 1 file(s) to review.

**Review by franciscojuniorharness — APPROVED** [review]
> Reviewed 1 of 8 files at r3, 3 of 8 files at r4, 1 of 2 files at r5, 3 of 3 files at r7, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45375)* status: 1 file(s) to review.

**Review by franciscojuniorharness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45375)* status: 1 file(s) to review.

**Review by franciscojuniorharness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45375#-:-NRJvZ_rDZ7cqH5c9Lse:bnfp4nl)* status: 1 file(s) to review.
